modules/ballerina-http/src/main/java/org/ballerinalang/net/http/HttpUtil.java

```diff
@@ -448,12 +454,41 @@ public static void startPendingHttpConnectors() throws BallerinaConnectorExcepti
         }
     }
 
-    public static void handleResponse(HTTPCarbonMessage requestMsg, HTTPCarbonMessage responseMsg) {
+    public static BValue[] prepareResponseAndSend(Context context, AbstractNativeFunction abstractNativeFunction
+            , HTTPCarbonMessage requestMessage, HTTPCarbonMessage responseMessage) {
+        addHTTPSessionAndCorsHeaders(requestMessage, responseMessage);
+        HttpResponseStatusFuture statusFuture = handleResponse(requestMessage, responseMessage);
         try {
-            requestMsg.respond(responseMsg);
+            statusFuture = statusFuture.sync();
+        } catch (InterruptedException e) {
+            throw new BallerinaException("interrupted sync: " + e.getMessage());
+        }
+        if (statusFuture.getStatus().getCause() != null) {
+            return abstractNativeFunction.getBValues(getServerConnectorError(context
+                    , statusFuture.getStatus().getCause()));
+        }
+        return abstractNativeFunction.VOID_RETURN;
+    }
+
+    public static void addHTTPSessionAndCorsHeaders(HTTPCarbonMessage requestMsg, HTTPCarbonMessage responseMsg) {
+        Session session = (Session) requestMsg.getProperty(Constants.HTTP_SESSION);
+        if (session != null) {
+            session.generateSessionHeader(responseMsg);
+        }
+        //Process CORS if exists.
+        if (requestMsg.getHeader("Origin") != null) {
```

Comment:
Shall we use a Constant defined in org.ballerinalang.net.http.Constants for "Origin" here?

Reply:
Ack